### PR TITLE
Host Validation Doc

### PR DESCRIPTION
**Pull request by aric49**
A document outlining proposed host validation tests.

### CODE REVIEW

File: docs/Host_Validation_Architecture.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+1. __Generate SHA256 hashes of all Promenade Generated configuration files and certificates:__ This process will run during the initial deployment of all the Kubernetes hosts in the cluster. Promenade will write a file to `/var/log/promenade/promenade_file_hash.log` with the path to the file as well as the SHA256 hash of that file when it was originally generated. Subsequent runs of the host validation script will re-calculate the SHA256 hashes of the files as they are presently, and compare them with the values written in the log file at the time of deployment. If the file hashes differ, the host validation script will log a failure and provide the user with a list of files that have potentially been modified.
+
+2. __Issue a REST Post Request to the Kubernetes API:__ This test will trigger a simple POST request to the internernal and external Kubernetes API and validate that it gets a `200` response.  This test will check to see if the Kubernetes API is running and responding to requests as needed.
```
Comment (gardlt) [inline docs/Host_Validation_Architecture.md]: internal?

**Comment by gardlt** [comment]
Is this suppose to be a feature enhancement proposal?

**Comment by aric49** [comment]
@gardlt Kinda --- this is from a conversation @mark-burnett and I had yesterday regarding the best way to tackle smoke testing to validate that hosts come up properly during a Promenade Kubernetes deployment.  The goal of this document is to provide insight into what will be in scope for Promenade vs other tools like Armada, Prometheus, or other configuration management platforms which may run in conjunction with Promenade. The idea of this document is to get the conversation started and iterate on it.

File: docs/Host_Validation_Architecture.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Host Validation and Testing
+
+__Date:__  7/6/2017
+
+__Objective:__ To determine the best strategy to validate hosts have joined the Kubernetes cluster successfully and are ready to schedule workloads
+
+## Problem Statement
+The Promenade project is a simple and efficient tool for orchestrating the deployment of self-hosted Kubernetes clusters.  However, Promenade currently lacks the ability to validate that clusters have been deployed successfully and are actually ready to schedule workloads.   I would like to propose a series of tests and validations that can be included with Promenade which will ensure the Kubernetes clusters are working as intended at the time of deployment, and can be executed at any point in time  after the cluster has been deployed to ensure it is continuing to function as intended for Day 2 operations. This tool should be able to identify potential problems with the Kubernetes cluster and provide the user with a comprehensive debugging output which highlights the identified issues.
```
Comment (intlabs) [inline docs/Host_Validation_Architecture.md]: Is promenade self-hosted?

Nit: spacing.

File: docs/Host_Validation_Architecture.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Identified Tests
+I have identified the following tests which I believe to be a good base-line for basic Kubernetes smoke testing at the time of cluster deployment and during cluster life-cycle management:
+
+1. __Generate SHA256 hashes of all Promenade Generated configuration files and certificates:__ This process will run during the initial deployment of all the Kubernetes hosts in the cluster. Promenade will write a file to `/var/log/promenade/promenade_file_hash.log` with the path to the file as well as the SHA256 hash of that file when it was originally generated. Subsequent runs of the host validation script will re-calculate the SHA256 hashes of the files as they are presently, and compare them with the values written in the log file at the time of deployment. If the file hashes differ, the host validation script will log a failure and provide the user with a list of files that have potentially been modified.
```
Comment (intlabs) [inline docs/Host_Validation_Architecture.md]: What is the purpose of this test?

File: docs/Host_Validation_Architecture.md (hunk 1 above)
Comment (intlabs) [reply docs/Host_Validation_Architecture.md]: would it not also make sense to simply check that the k8s API is reporting all pods in the `kube-system` namespace as running and in a healthy state

File: docs/Host_Validation_Architecture.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+2. __Issue a REST Post Request to the Kubernetes API:__ This test will trigger a simple POST request to the internernal and external Kubernetes API and validate that it gets a `200` response.  This test will check to see if the Kubernetes API is running and responding to requests as needed.
+
+3. __Kubelet and Docker Validation__ : The goal of this test will be to ensure the Kubelet and Docker are present on the host and running without error.
+
```
Comment (intlabs) [inline docs/Host_Validation_Architecture.md]: It would make sense to monitor the etcd cluster, in addition to general health checks it would be prudent to monitor the logs for potential I/O issues and timeouts for actions.

File: docs/Host_Validation_Architecture.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+## Other Concerns
+
+1. __Monitoring Tools vs. Promenade:__ A lot of the identified tests blur the lines between what the monitoring tools such as Prometheus are designed to do.  We don't want two tools performing re-work, so we will need to clearly define the boundaries in the scope of Promenade versus other monitoring tools.
+
+2. __Configuration Management Tools vs. Promenade:__  Much of what Promenade does today is template and place configuration files using Python and Jinja2 templates. Purpose-built configuration management tools such as Ansible, Chef, and Puppet can do all of these things, on top of validations to ensure the files are in the desired state.  Instead of re-inventing the wheel, would it make more sense to leverage a configuration management tool and build a wrapper around it for things such as validating configuration files and ensuring services are running?  Should we create a separate project which Promenade can leverage that is built for configuration management using one of the already existing platforms? This conversation again goes back to cleanly defining the scope of Promenade vs other tools that we plan on using going forward.
```
Comment (intlabs) [inline docs/Host_Validation_Architecture.md:27]: I think this approach makes good sense - it should be possible for the activities scoped here to be performed by a lightweight tool that worked regardless of the deployment methodology of the k8s cluster.

**Review by intlabs — CHANGES_REQUESTED** [review]

File: docs/Host_Validation_Architecture.md (hunk 3 above)
Comment (aric49) [reply docs/Host_Validation_Architecture.md]: The purpose of this test would be to check to see if the files generated by Promenade have been modified in the time since the deployment. If the SHA256 hashes created at the time of deployment do not line up with the SHA256 hashes at the time the test is run, then a change has been made to that file which may (or may not) impact the functionality of the K8s cluster. The purpose of this test is to provide a list of files that have changed since deployment to help with future debugging.

File: docs/Host_Validation_Architecture.md (hunk 4 above)
Comment (aric49) [reply docs/Host_Validation_Architecture.md]: Good idea.  I completely agree.  That would give us great insight into the overall cluster health --- however, I feel as though more complicated tests like that are better suited for purpose-built monitoring tools (feel free to disagree).     I think the purpose of this document is to provide clear scoping as to what the expectations of Promenade are vs these other tools so we aren't re-inventing the wheel as such.    I can be convinced this would fall into the scope of Promenade, but I think we need to be super clear where the boundaries are.   I think that Promenade should do simple smoke testing to ensure cluster functionality, but would we want to evolve Promenade into a fully functional monitoring and config management platform? Or at that point, is it better to leverage other tools?  The questions I am posing in this document are what should those tests and configuration management tasks should be from the perspective of Promenade.

File: docs/Host_Validation_Architecture.md (hunk 1 above)
Comment (aric49) [reply docs/Host_Validation_Architecture.md]: I agree -- that'd be a better test as it would ensure the functionality of the API as well as the pods. I'll make that change.

File: docs/Host_Validation_Architecture.md (hunk 4 above)
Comment (wilkers-steve) [reply docs/Host_Validation_Architecture.md]: With respect to evolving Promenade into a monitoring and configuration management platform -- absolutely not.  However, monitoring etcd does make sense here I think.  If the goal here is to validate the cluster is functional, making sure etcd is handling requests properly and not experiencing timeouts is crucial.  I don't feel this is unreasonable as it relates to the proposed scope here.

File: docs/Host_Validation_Architecture.md (hunk 5 above)
Comment (wilkers-steve) [reply docs/Host_Validation_Architecture.md:27]: My concern is that all points raised here expand the scope of Promenade.  I would much prefer see something lightweight and independent with the sole purpose of validating a cluster period.  I think coupling this tightly to Promenade is the wrong approach.

**Review by wilkers-steve — COMMENTED** [review]
Some comments inline.

The tl;dr is that I feel expanding Promenade's scope to handle validation testing and some monitoring should be reconsidered.  I'd much rather see something else less coupled to Promenade that can validate a cluster regardless of whether Promenade deployed it or it's launched by other means.  Such a tool would add value beyond this repository

File: docs/Host_Validation_Architecture.md (hunk 4 above)
Comment (aric49) [reply docs/Host_Validation_Architecture.md]: I think that makes total sense @wilkers-steve --- My concern was incorporating a periodic health check of the etcd service kind of starts blending into the sphere of monitoring.   I think a script that kicks off and checks the health of etcd during deployment, and again if the user triggers one is super reasonable. Is that what you are referring to?

File: docs/Host_Validation_Architecture.md (hunk 3 above)
Comment (intlabs) [reply docs/Host_Validation_Architecture.md]: would run a simple `find` command not achieve the same thing? If a historical log of configs is required would it not make sense for them to be checked into a local git repo?

File: docs/Host_Validation_Architecture.md (hunk 5 above)
Comment (alanmeadows) [reply docs/Host_Validation_Architecture.md:27]: The high level goal is indeed cluster validation - the requirement is that Promenade call it, and I would imagine, needs to inform it about its view of how the world should be to allow the tool to intelligently identify gaps with that world view, but that does not necessarily require that the tool be contained within promenade.  

We could either split this out now, or evolve it to its own project once it takes shape.  I'm fine with either.

Just a note -- cluster validation at a general level for example doing something such as etcd cluster validation, and checking its up, it has a quorum, and its healthy is useful but we want to ultimately inform this tool about knowledge only Promenade has.  If its external, promenade will simply need to pass the relevant data along. For instance, in the etcd scenario, the cluster being healthy does not necessarily mean that the cluster member count is identical to the number of hosts targeted to receive said etcd workload which is solely within Promenades own knowledge.   If an external tool, which makes sense on several levels, there will need to be a path to inform it of the intended configuration so it can validate it has been achieved.

File: docs/Host_Validation_Architecture.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+2. __Issue a REST Post Request to the Kubernetes API:__ This test will trigger a simple POST request to the internernal and external Kubernetes API and validate that it gets a `200` response. This check wil also query the `kube-system` namespace to ensure all pods are in a `RUNNING` state. This will ensure that the K8's API and system pods are running and responding as expected. 
+
+3. __Kubelet and Docker Validation__ : The goal of this test will be to ensure the Kubelet and Docker are present on the host and running without error.
```
Comment (alanmeadows) [inline docs/Host_Validation_Architecture.md:19]: Since promenade knows the target docker version (and Kubernetes version), would we not validate we've achieved that and something else didn't manage to muck with that either during install or post install?

File: docs/Host_Validation_Architecture.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+4. __Test Workloads__:  This will be the final test which is run against the Kubernetes cluster. If the other tests identified above pass without error, Promenade will issue a test deployment of pods on the cluster and ensure they come up in the `RUNNING` state with the correct IPs and networking provided by the CNI.  Once the workload has come up successfully, Promenade will destroy the deployment and ensure the pods are terminated cleanly.  This final test will determine if the Kubernetes cluster nodes can successfully schedule workloads.
```
Comment (alanmeadows) [inline docs/Host_Validation_Architecture.md:21]: We should be clear the goal is to test that *every node* in the cluster can receive a pod deployment successfully.  We may need to describe the various kinds of deployments necessary to achieve that.  For instance, I expect a CNI based pod (not hostNetwork) needs to be spawned to validate CNI functionality.  Are there any other scenarios that should be tested?

We should also validate pod east-west functionality -- my ideal scenario and complex scenario is a pod on every single node, a full mesh connectivity test between every pod to every node indicating the site will not experience any oddities with container workload reachability.

File: docs/Host_Validation_Architecture.md (hunk 5 above)
Comment (alanmeadows) [reply docs/Host_Validation_Architecture.md:27]: The only requirement is that promenade call this functionality and provide it the details it needs to perform its validation.

The need here is for this function to be nicely contained in a container and I would hope running in Kubernetes itself. 

Finally, I would not too quickly point the compass at configuration management tools - they may provide the functionality needed for file state validation (usually geared toward them managing said file which is not the case here), but most of the validation covered outside of that is geared more toward interrogating etcd, kubernetes (as an API client), and potentially other APIs.  It may be another case for .py.

**Comment by alanmeadows** [comment]
One further thought.

We should evaluate some of the code in the Kubernetes CI/CD gating system for this purpose.  We obviously know that in addition to very specific code paths and complex testing to ensure they haven't broken edge case code, they also likely have tests for some of the high level elements we're attempting to validate here.

We should explore what they have and how easy it would be benefit our self-contained version based on tests they've already outlined.

At the very least, much of what they test can educate us on what we need to evaluate. For example, this document isn't low level enough to suggest, but eventually I expect this component (or external project) to articulate use cases along the lines of:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/kubelet_test.go#L40-L71

- Create a pod, retrieve its logs, validate they have what we expect.

I am not suggesting that we recreate all of the integration tests Kubernetes already does, but these tests like this help validate basic operations with the full stack (our k8s, our docker version, our operating system configuration, where we are storing docker logs, etc.) in running environments and most importantly for us in these larger environments _on a per node basis_ to check for abhorrent behavior earlier rather then later.

File: docs/Host_Validation_Architecture.md (hunk 3 above)
Comment (v1k0d3n) [reply docs/Host_Validation_Architecture.md]: this is just a nit, but if its just for checksum why use SHA256 vs. something more expeditious like MD5?

Comment (intlabs) [reply docs/Host_Validation_Architecture.md]: @v1k0d3n your point is spot on in regard to the right algo for the job, but unfortunatly most code-analysis tools flag MD5 as insecure (le sigh) - therefore the extra couple of cyles spent hashing a config file seems inconsquential to the number of emails later. That said I'm still not sure what this method achives over running a `find` command - esp as looking at the WIP current code proposed you would be able to bypass this mechanism by simply deleteing/updating the  `/var/log/promenade/promenade_file_hash.log` file (and when you think about file perms, preventing this senario becomes difficult - as presumably the k8s configs should only be writeable by root ). So I think it needs to be specified at the outset if this is a security feature, or a sanity check - and if its the latter why it employs sich a complex mechanism for performing this this task.

**Comment by mark-burnett** [comment]
@intlabs @wilkers-steve @aric49 @alanmeadows @v1k0d3n 

CC: @bryan-strassner

I agree with a lot of the good points made here, especially regarding the focus of responsibility of particular services/code.

It feels like the three main categories of validation that have come up are:

1. Host configuration validation
    - Presence/content of particular files on nodes
    - Versions of installed packages (in particular docker since Promenade is currently installing that directly, as called out by @alanmeadows)
2. Overall cluster validation
    - Exercising the Kubernetes API/use some Kubernetes integration tests
    - Reachability among pods running on all nodes to one another
    - Ability to allocate cluster resources to all (appropriate) nodes, e.g. PVs
3. Critical application validation
    - Primarily etcd cluster health?
    - Other applications Drydock, MaaS, Ceph, Airflow, etc.?

It actually seems that these categories could each reasonably be delegated to different components.  There's no reason to really expect a general cluster validation tool to check packages and files on nodes.  Likewise, while tools like ansible are pretty good at doing tasks in category 1, I don't really love the idea of writing an ansible module to check etcd health/configuration.

To me the questions become:

1. How should we prioritize these different aspects of validation?  It seems to me that 2. is more likely to expose serious problems and probably should be implemented first.  On the other hand, the story @aric49 was working on seemed to be more about 1 (I might be mis-interpreting the intent).
2. How much should we worry about separating these concerns now vs. later?
3. Will other applications need to initiate these various kinds of checks besides Promenade at cluster formation?

**Comment by mark-burnett** [comment]
Closing as we have moved to gerrithub.
